src/main/scala/cromwell/binding/types/WdlType.scala

```diff
@@ -57,11 +59,27 @@ trait WdlType {
 object WdlType {
   val parser = new WdlParser()
 
-  private lazy val wdlTypes = Seq(
-    WdlBooleanType, WdlExpressionType, WdlFileType, WdlFloatType,
-    WdlIntegerType, WdlNamespaceType, WdlObjectType, WdlStringType
+  /* This is in the order of coercion from non-wdl types */
+  val wdlTypes: Seq[WdlType] = Seq(
+    WdlStringType, WdlIntegerType, WdlFloatType, WdlMapType(WdlAnyType, WdlAnyType),
+    WdlArrayType(WdlAnyType), WdlBooleanType, WdlObjectType,
+
+    /* Nothing should coerce to these.  Anything that WdlFileType
+     * would match would be caught by WdlStringType first.  Further
+     * coercion can happen to convert WdlStringType -> WdlFileType.
+     */
+    WdlExpressionType, WdlFileType, WdlNamespaceType
   )
 
+  def areTypesHomogeneous(values: Iterable[WdlValue]): Try[WdlType] = {
+    values.map(_.wdlType).toSet match {
```

Comment:
If `values.map(_.wdlType)` evaluates to `Seq(WdlInteger, WdlInteger, WdlInteger)`, then the set would be `Set(WdlInteger)` and if the types are homogeneous then this set is of size 1.

I don't know how to use "distinct"